### ISSUE TITLE
Request for Clarification: reproducing individual Macaque reaching sessions using example notebook

### ISSUE THREAD

**Issue by XiaoyuZeng** [Dynamics-of-Neural-Systems-Lab/MARBLE#37]
Dear MARBLE Developers,

First, thank you again for creating and maintaining this excellent toolkit. It's a very valuable resource for the community, and I imagine it will become increasingly popular in the field. I'm planning to apply this method to my own single-neuron data from the human brain once I'm confident that I'm following the best practices (workflow, parameters, etc.).

I am currently working on replicating the results for Session 6 of macaque reaching task from your paper. I have successfully run the example notebook workflow_and_comparison_of_MARBLE_and_CEBRA.ipynb, and the code completes without errors now.

However, when I compare my generated results to the corresponding figure in the paper, I'm noticing that the outputs for some experimental conditions are significantly different.

My question is: **Is this level of difference expected due to the stochastic nature of the model fitting process (i.e., random noise), or is it more likely that there's a difference in the parameters I'm using compared to those used for the paper?**

For your reference, I have attached a side-by-side comparison showing the figure from the paper and the plot I generated.

<img width="775" height="379" alt="Image" src="https://github.com/user-attachments/assets/f44dec5f-37c2-4bd6-8f27-5fc4b25fcb38" />

I have also listed the key data and parameters used in my reproduction below:
- data/rate_data_20ms_100ms.pkl
- pca_n = 5
- filter_data = True
- day = 6 #load one session
- data = MARBLE.construct_dataset(
    anchor=pos,
    vector=vel,
    k=30,
    spacing=0.0,
    delta=1.5,
)
- Define split sizes (e.g., 80% train, 10% val, 10% test)
train_size = int(0.8 * num_nodes)
val_size = int(0.1 * num_nodes)
- params = {
    "epochs":120,  # optimisation epochs
    "order": 2,  # order of derivatives
    "hidden_channels": 100,  # number of internal dimensions in MLP
    "out_channels": 3, 
    "inner_product_features": False,
    "diffusion": True,
}

One critical detail is that I had to generate the train (80%), validation (10%), and test (10%) masks myself, because the version of MARBLE.construct_dataset I used did not generate them. 

Any clarification you could provide would be greatly appreciated. If there are specific parameter settings that were used for the paper that might differ from the defaults in the example code, could you please point me in the right direction?

Thank you for your time and help.

Best regards,
Xiaoyu

**Comment by agosztolai** [comment]
Hi Xiaoyu,

Thanks for the positive feedback. 

The difference in output you observe is expected due to different seedings.

1) Global alignment. Two marble embeddings will be different up to a linear alignment. This is a direct consequence of contrastive learning and is mathematically provable. This is mostly what you are seeing: the results are differently oriented and reflected but the global structure is consistent.

2) Stochastic differences. Since you start training from different instantiations, the network will give slightly different results. 

So, the results you see are consistent with the method working. 

Unfortunately, it is not possible to exactly reproduce the outputs because the PyTorch Geometric function `NeighborSampler` cannot be seeded (at least this was the case for the version we used).

In general, a good practice is to train several marble networks for a given data. This way you are confident that you are getting consistent results. If you report metrics such as decoding, then evaluate this on all instances and report fluctuations.

Hope this helps.

**Comment by XiaoyuZeng** [comment]
> Hi Xiaoyu,
> 
> Thanks for the positive feedback.
> 
> The difference in output you observe is expected due to different seedings.
> 
> 1. Global alignment. Two marble embeddings will be different up to a linear alignment. This is a direct consequence of contrastive learning and is mathematically provable. This is mostly what you are seeing: the results are differently oriented and reflected but the global structure is consistent.
> 2. Stochastic differences. Since you start training from different instantiations, the network will give slightly different results.
> 
> So, the results you see are consistent with the method working.
> 
> Unfortunately, it is not possible to exactly reproduce the outputs because the PyTorch Geometric function `NeighborSampler` cannot be seeded (at least this was the case for the version we used).
> 
> In general, a good practice is to train several marble networks for a given data. This way you are confident that you are getting consistent results. If you report metrics such as decoding, then evaluate this on all instances and report fluctuations.
> 
> Hope this helps.

Thank you so much for your detailed reply! This is extremely helpful.

I understand now that the variance in the outputs is expected. Please feel free to close this issue.

Thanks again for your time and help!

Best,
Xiaoyu
